Add custom light and dark themes

phoerious opened the pull request:
[TIP]:  # ( Provide a general summary of your changes in the title above ^^ )

## Type of change
[NOTE]: # ( Please remove all lines which don't apply. )
- ✅ New feature (non-breaking change which adds functionality)

## Description and Context
[NOTE]: # ( Describe your changes in detail, why is this change required? )
[NOTE]: # ( Describe the context of your change. Explain large code modifications. )
[NOTE]: # ( If it fixes an open issue, please add "Fixes #XXX" as necessary )
Since the platform-native widgets are very hard to customise without breaking stuff and people keep (rightfully) complaining about KeePassXC's "oldfashioned" design, this PR adds a custom-made light and dark theme.

The themes are based on Andrew Richard's Phantomstyle (https://github.com/randrew/phantomstyle) with many modifications and custom CSS styling. Since they are implemented as QStyles, they require no UI rewrite and can be provided alongside the "classic" look for those who may prefer it.

## Screenshots
[TIP]:  # ( Do not include screenshots of your actual database! )

### Light:
![Screenshot 2020-01-18 at 17 16 03](https://user-images.githubusercontent.com/911270/72666853-53edb400-3a16-11ea-83b6-3f829655fd3c.png)
![image](https://user-images.githubusercontent.com/911270/72667204-95339300-3a19-11ea-9dd0-867e21691530.png)
![image](https://user-images.githubusercontent.com/911270/72667218-c3b16e00-3a19-11ea-901c-4405cf22299a.png)

### Dark:

![image](https://user-images.githubusercontent.com/911270/75611315-cd9ab680-5b19-11ea-9548-acdee30c474d.png)


## Checklist:
[NOTE]: # ( Please go over all the following points. )
[NOTE]: # ( Again, remove any lines which don't apply. )
[NOTE]: # ( Pull Requests that don't fulfill all [REQUIRED] requisites are likely )
[NOTE]: # ( to be sent back to you for correction or will be rejected.  )
- ✅ I have read the **CONTRIBUTING** document. **[REQUIRED]**
- ✅ My code follows the code style of this project. **[REQUIRED]**
- ✅ All new and existing tests passed. **[REQUIRED]**
- ✅ I have compiled and verified my code with `-DWITH_ASAN=ON`. **[REQUIRED]**

droidmonkey commented:
I don't like the centered tabs or section labels

phoerious commented:
Yeah, we can discuss that. It's something I wanted to try, but I'm not quite sure yet if I'm satisfied with it. There are also other things I don't yet like such as the buttons in the light theme. The colour shades also need to be tuned for better harmony.

droidmonkey commented:
The search help panel needs to be written in HTML + CSS

varjolintu commented:
Here's how it looks on non-HiDPI display with macOS High Sierra, which doesn't have a native dark theme support like Mojave:

Light theme:
![light](https://user-images.githubusercontent.com/24570482/71955675-9ed72280-31f1-11ea-99fe-2fb060048aff.png)

Dark theme:
![dark](https://user-images.githubusercontent.com/24570482/71955680-a4346d00-31f1-11ea-9a9e-b32c991c6632.png)

"Native" theme:
![native](https://user-images.githubusercontent.com/24570482/71955687-a991b780-31f1-11ea-84c4-fcdafc3eedfa.png)

Current native theme with 2.5.2 for comparison:
![current_native](https://user-images.githubusercontent.com/24570482/71955698-b1e9f280-31f1-11ea-833a-764f35341e31.png)

phoerious commented:
Did you restart KeePassXC after switching to native? On KDE it looked more or less okay, but it's possible that reapplying the proxy style doesn't always work properly.

varjolintu commented:
@phoerious Restarting fixed it. However, every time you change the theme for the first time, it takes around 5-10 seconds on my machine and KeePassXC hits 100% CPU. Plus the warning message for dev build still has white text when dark theme is enabled.

phoerious commented:
I did not fix the message widget yet, but I am aware of it.
On Linux I could change the theme without issues. It only took a second or so.

OLLI-S commented:
I really love the new themes, this makes KeePassXC more modern.

In the 3rd screen shot (settings) I see that the headlines of the groups (in the old layout you had group boxes) are centered.
Under usability issues / software-ergonomics issues this is problematic, because you have a long distance between the headline and the fields below the headlines:

![image](https://user-images.githubusercontent.com/8037790/72019003-1300dd00-3269-11ea-80b0-ab8ab694a010.png)

If the user opens the settings then he focuses elements that are on the left side, because this is a learned layout from other applications (we also read from left to right). So the first element that the user recognizes is the list of checkboxes.

But in what context are these checkboxes?
There is no headline shown above the checkboxes (the headline is not left-aligned) so it is not recognized at once).
This irritates the user and this is a little stop in the workflow, because he has to search any information what is the context of the checkboxes.

This searching requires user-action (he has to move the eyes and also the read to search for the headline.
This costs a little bit time, not much but it interrupts the process.

After the user has read the headline, he has to focus the checkboxes again (also a larger distance).

If you show the headlines left-aligned they are recognized much faster and the user does not have to search for them.
This is saves the user some time, does not interrupt the process and this layout (headlines left-aligned) is a common and learned layout).

Same for the tabs, they should also be left-aligned.

Here speaks my "usability" side (I work in usability, ux, accessibility and software-ergonomics.

Aetf commented:
How does this integrate with native Qt themes, like KDE Breeze?

4jNsY6fCVqZv commented:
@phoerious Thank you for your work!

What I notice in the settings view:

1. There's an incredible amount of white space. Unlike in 2.3.1.
Can you find a more appealing solution for this? 
2. Concerning the selection controls: Can you please look if other forms would make more sense than checkboxes? For inspiration take a look at: https://elementary.io/docs/human-interface-guidelines#selection-controls
3. In general to the settings: The category is called "Basic settings". Would it be possible to simplify the content here again? There are simply a lot of settings. So either split the window between "Basic settings" and "Advanced settings" or adjust the page performance tabs accordingly? For example the SSH Agent could run well under an advanced category. Also, the browser integration and security and general tab could fit into both categories and simplify the user experience. What are your thoughts on this?

phoerious commented:
@OLLI-S Centred tabs is something I wanted to try, since it looks familiar for macOS users (and macOS users deserve a better-looking UI the most I suppose). I do agree that it doesn't really work here and will change it back to left-aligned tabs. There are also a few other things I have changed already which haven't been pushed yet (such as getting rid of rounded corners entirely). Thanks for your analysis, it is much appreciated. If you find anything else, please do tell.

@Aetf It doesn't, but for KDE, you can always stick with the "classic" theme. I will integrate some platform-specific styles, though, for instance to match the background colour with the window title colour on macOS. We already have some macOS-specific alignment fixes for the "classic" theme, particularly for buttons, which I need to adjust as well (see the weird layout bugs above).

Someone else also wrote something here, but I can't find the comment anymore. The fact that the new themes contain more white space is by design (punny, I know) as the previous theme was very cramped. Perhaps I can make that configurable, but for now it's there to stay.

OLLI-S commented:
@phoerious If you have changed the layout so the posted screen shots are no longer up-to-date, can you please post new ones?

I showed the screen shots above to a colleague of mine.
He uses Apple devices (also iPhone and Apple Watch) and also has the newest Beta of the OS on all devices.
He also is a designer.

His feedback after a quick look is that the layout is "old fashioned" (it was used many years ago) and not state-of-the-art.
I will investigate him for deeper feedback.
His first

phoerious commented:
The layout is a pretty standard GUI layout, not a fancy web UI. I would like to keep these two things separate. This PR is only about the theme, not about the GUI layout, which we can only change to a certain degree anyway due to the underlying widget toolkit.

droidmonkey commented:
What's old is new again :stuck_out_tongue_winking_eye:

phoerious commented:
Yeah, I'm not against redesigning individual widgets, but I would like to keep the extra effort in balance with the gains.

OLLI-S commented:
Do you have to be a programmer to design the widgets?
Or is there an "editor" that you can use without programming skills?

phoerious commented:
There is an editor, but it is extremely technical and not good for anything beyond stitching together standard widgets.

grenzor commented:
Looks nice @phoerious

Is there a possibility for users to be able use these themes as templates to edit and make their own themes within the application or is that not possible?

phoerious commented:
It certainly is doable, but I don't like the idea of users loading arbitrary themes into their password manager from a security standpoint. Nor do I see the necessity if there is a good default theme to begin with.

grenzor commented:
I can see people liking the ability to edit themes according to their own preferences. If there is a security issue then I guess that takes priority (though I'm not sure what the severity of loading custom themes is)

droidmonkey commented:
Themes include both style (CSS) and functionality/looks (C++ plugin). We can certainly support custom style sheet loading, but not themes.

phoerious commented:
You can already set a custom application stylesheet via the `-stylesheet` and option and load a custom style plugin via the `-style` option. The former may be safe (but cannot replace a full theme), the latter involves loading arbitrary shared libraries.

wolframroesler commented:
@phoerious in #4066:

> @wolframroesler How do we handle dark themes and selected states of icons? In my PR #4110 I have an ugly intermediary solution where I convert the SVGs to 128x128 pixmaps and then recolour them a) for inverted icons in selection highlights and b) for dark themes (which I try to detect based on the standard palette).
> 
> I think there might be a more elegant way to achieve this which does not include creating fixed-size pixmaps nor the need for two different icon sets.
> 
> Also an interesting observation: Providing only 128x128 icons and then letting Qt do the rest works better than creating multiple sizes beforehand. Perhaps that's because some icons are not exactly 16x16 or so and scaling them to the correct size from 128 produces better results than scaling from the nearest already-low resolution, who knows.

Changing the color of an SVG from black to white should be possible with something like

```sh
$ sed 's/path /path style="fill:#FFFFFF;" /g' <icon.svg >white.svg
```

We could easily batch-run this over the existing SVGs and then ship two versions (one black and one white) of each. Haven't tried inverting (for selected icons) but maybe it's equally easy. So perhaps we'll need four versions of each SVG (black regular, black selected, white regular, white selected).

For clarification: Do you mean pre-generating 128x128 PNGs gives better results than the PNGs pre-generated in smaller sizes, or better results than the SVGs?

phoerious commented:
> Changing the color of an SVG from black to white should be possible with something like

Yeah, I'm not at all worried about that. My concern is that I want to avoid this:

>  So perhaps we'll need four versions of each SVG

I don't really want to bundle four largely redundant icon sets, because it blows up the size of our binaries quite significantly, especially if those icons are PNGs and perhaps even at multiple resolutions.

Generating the icons on demand has the advantage of saving space and allowing us to adjust the icons to the theme's palette (instead of just plain black and white or so). The downside, on the other hand, is the computational overhead. I want to find a good solution for this. I'm also not sure if we should generate PNGs or SVGs on demand. I would assume the latter saves a bit of memory, but at the cost of having to parse and render an SVG twice. I also don't really know how you would load an in-memory SVG into a QIcon without writing it into a QBuffer and reading that as a file, which sounds silly to me (perhaps there is a more elegant way).

> For clarification: Do you mean pre-generating 128x128 PNGs gives better results than the PNGs pre-generated in smaller sizes, or better results than the SVGs?

Generating and adding only a 128px resolution to any icon gave better results than generating and adding multiple pre-computed resolutions.

lindhe commented:
I haven't compiled to check if @varjolintu's  image above is out-of-date, so perhaps this is already fixed. But let's not have dark text on dark background in the tab titles, yeah?

> Dark theme:
> ![dark](https://user-images.githubusercontent.com/24570482/71955680-a4346d00-31f1-11ea-9a9e-b32c991c6632.png)

phoerious commented:
Your screenshot is outdated, please update your branch. In any case, the dark theme is still not finished. I am getting the light theme up to speed on all platforms first (see updated screenshot above). The dark palette is also not final, too much green in the dark shades and too little contrast overall.

droidmonkey commented:
Issues:

1. Close button hover adds unnecessary background color:
![image](https://user-images.githubusercontent.com/2809491/75594074-7c28f380-5a55-11ea-9858-c9c438e96186.png)

2. Pixel perfection issues on tabs (may not be correctable):
![image](https://user-images.githubusercontent.com/2809491/75594163-d4f88c00-5a55-11ea-8a26-d649b5c85820.png)

3. Pixel perfection issues on welcome screen:
![image](https://user-images.githubusercontent.com/2809491/75594358-7089fc80-5a56-11ea-97cc-07de9562fe3c.png)

4. Bad colors for password strength and very hard to see length slider slug:
![image](https://user-images.githubusercontent.com/2809491/75594408-a0d19b00-5a56-11ea-9407-aab7f018d753.png)

5. Recommend making section headers bold font and adding top margin to distinguish them from each other:
![image](https://user-images.githubusercontent.com/2809491/75594530-0887e600-5a57-11ea-8242-fdd3ae90e86d.png)

phoerious commented:
I though I had pixel-perfected the controls already, but seems like something broke in between. I fixed that; all controls now have the same height (buttons, line edits, spinboxes, comboboxes).

The blue highlight is a Windows-specific quirk. The highlight should actually be green, but somehow Windows ignores that. I guess it's fixable somehow, but it's not directly a theming issue I suppose. More like a downstream bug in the widget.

The link colours seems unfixable. I added disabled link colours, but at least on Linux, Qt still uses the normal link colour.

Please check my recent commits. I also adjusted the dark theme colours to improve contrast.

droidmonkey commented:
OK perfect. Still need to address the password length slug and the blue color if at all possible (perhaps force a style class on these elements?).

Also some more findings:

1. Database settings left bar is very wide:
![image](https://user-images.githubusercontent.com/2809491/75609755-357cdd00-5ad9-11ea-8f30-10216b1d9ffd.png)

2. There should be a distinction between combo boxes that are paired with buttons (taller) and those paired with settings text (shorter):
![image](https://user-images.githubusercontent.com/2809491/75609771-4cbbca80-5ad9-11ea-8d56-e5e4910fb27f.png)

We really need to ditch these default database icons...
![image](https://user-images.githubusercontent.com/2809491/75609741-1e3def80-5ad9-11ea-9d33-ffcf2b27c4b6.png)

phoerious commented:
> Still need to address the password length slug and the blue color 

I increased the contrast as much as possible. The blue colour is probably out of scope of this PR. I assume it needs to be fixed in the sidebar widget.

> Database settings left bar is very wide

That's not a theme bug and has been there before. Also needs to be fixed in the widget.

> There should be a distinction between combo boxes that are paired with buttons (taller) and those paired with settings text (shorter):

Not sure what you mean. The length depends on whatever you define for this particular widget. The whole settings page needs to be overhauled, but that's again a different issue.

droidmonkey commented:
Not the length, the height. Inline combo boxes should be only slightly taller then the text.

phoerious commented:
That's not possible. They can only have one uniform height.

OLLI-S commented:
Where will users be able to select the theme?
I assume in the settings?

phoerious commented:
Yes.

varjolintu commented:
Is the active page indicator showing a green text background the indended behavior? Looking at the previous screenshots the whole icon & text are with a coloured background. This happens with macOS High Sierra and Mojave:

![Screen Shot 2020-03-01 at 15 08 49](https://user-images.githubusercontent.com/24570482/75626394-ebc2ee00-5bcf-11ea-9622-32da58782611.png)

Also, this branch doesn't remember any Browser Integration settings that has been previously set, or reset. All settings are forgotten if the application is restarted.

phoerious commented:
It's a known macOS bug, but none that I could fix on theme side. The widget needs sie updates.

droidmonkey commented:
I've noticed that when I enable browser integration on my debug build it messes with the checkbox settings on my release install. Anecdotal observation, haven't proven it yet, but suspicious.

phoerious commented:
Interesting, although I don't know why it would do that..

varjolintu commented:
Same happens under Linux when I debug via Qt Creator. I quickly debugged it and I guess it has something to do with reading the correct settings. When enabling the Browser Integration support again, `Config::set` returns because it has been already set to enabled. I'm investigating it some more.

varjolintu commented on src/gui/MainWindow.cpp:
```diff
@@ -911,7 +926,6 @@ void MainWindow::switchToDatabases()
 void MainWindow::switchToSettings(bool enabled)
 {
     if (enabled) {
-        m_ui->settingsWidget->loadSettings();
```
@phoerious This line removal breaks the Browser Integration settings.

phoerious replied:
Alright, good find. I remember I had some weird issues with settings not being loaded which I fixed, though. I will go through the diff and see if there are any incidental changes that shouldn't be there.

phoerious replied:
Fixed.

phoerious commented:
I think I am satisfied with this as a first mergeable version of this PR.

@wolframroesler Perhaps a TODO for you: We need a new binary mask version of the secret service icon. The current one cannot be recoloured:

![image](https://user-images.githubusercontent.com/911270/75767873-cb1ba500-5d43-11ea-85ae-fffa4ef39df3.png)

wolframroesler commented:
Requested the freedesktop.org icon (which is what we display for the Secret Service pane) from the simpleicons repo (https://github.com/simple-icons/simple-icons/issues/2727).

phoerious commented:
Awesome, thanks.

droidmonkey commented on src/gui/MainWindow.cpp:
```diff
@@ -618,14 +622,18 @@ void MainWindow::setMenuActionState(DatabaseWidget::Mode mode)
+        m_ui->toolbarSeparator->setVisible(
+            (!inWelcomeWidget && inDatabaseTabWidget && !m_ui->tabWidget->tabBar()->isVisible())
+            || currentIndex == SettingsScreen);
+    }
```
Should we really be "adding" this UI element, I feel like it just creates a double line that is not necessary:

![image](https://user-images.githubusercontent.com/2809491/75907887-223e7a00-5e17-11ea-8e26-37bd29284545.png)

phoerious replied:
It's needed for consistency. The tabbar creates the same line, but when the tabbar is hidden, the line vanishes. That destroys the illusion that the toolbar area is one homogeneous area  with the tabs in front. Without the line, the top part also bleeds too much into the lower part which is particularly irritating in the light theme and on macOS where the top part is visually very distinct from the rest of the application.

droidmonkey replied:
I agree it is necessary in light mode, but not in dark mode.

phoerious replied:
I think it is needed, but it should be darker than the toolbar background, not lighter.

OLLI-S replied:
I would prefer to see the line always, a clear visual border between the elements.

phoerious replied:
Although now I think about it, the bright line makes more sense. In fact, the light theme should have a bright line as well so that the toolbar area looks sunken not raised. Otherwise the tabs couldn't be in front. That would also be more consistent with the macOS look.

droidmonkey replied:
Hah, I think the separator looks good in light mode.

phoerious replied:
I changed the swatch colour from `S_window_outline` to `S_frame_outline`, which is the same as the tabbar border for light, but slightly different for dark. Then I noticed that despite the HLine shadow being `QFrame::Plain`, there was still a subtle emboss effect that simply wasn't visible in the light theme, but very much so on a dark background.

My fix was to adjust the HLine rendering to resemble that of the tabbar base where I draw a 2px border with a specular color and then draw the actual 1px frame over that. Now it looks basically like a solid line and not like a double line anymore. It changes the overall appearance, though. The bright line colour now makes the area below the toolbar look clearly raised on Windows, while it looks sunken with a bright background.

droidmonkey commented on src/gui/MainWindow.cpp:
```diff
@@ -776,6 +784,13 @@ void MainWindow::setMenuActionState(DatabaseWidget::Mode mode)
+    if (m_showToolbarSeparator) {
+        m_ui->toolbarSeparator->setVisible(!tabsVisible && m_ui->stackedWidget->currentIndex() == DatabaseTabScreen);
+    }
+}
```
See above comment

droidmonkey commented:
This looks fantastic, I recommend merging!

michaelblyons commented:
Can we see post-merge screenshots to get excited about? 😉

phoerious commented:
You can certainly build and test it.

droidmonkey commented:
Even easier! Download the snapshot: https://snapshot.keepassxc.org

wolframroesler commented:
@phoerious here's the freedesktop.org icon, fresh from the simple icons repo: https://github.com/PeterShaggyNoble/simple-icons/blob/c742dce6affe0d24f30376f17f97668fd4559dae/icons/freedesktop-dot-org.svg

wolframroesler commented:
The new theme looks really cool :) Great work @phoerious.

However, may I suggest you change the default theme to "platform native" for macOS builds? Most Mac people want all apps to look alike, and application-specific themes are usually frowned upon in the Mac world. It's OK to support the theme as an option, but it shouldn't be the default.

phoerious commented:
No, absolutely not. The classic theme is horrendous and far from what a platform-native look should be on macOS. Qt is doing a very poor job at supporting macOS properly. Also, there are probably more custom themes these days than actual "native" ones, so the whole "frowned upon" thing is more folklore than anything else.

phoerious commented:
That said, it's more important to follow the platform HIGs and provide an overall pleasant experience than to use native widgets. Qt itself certainly does neither. And macOS users from my experience are usually quite content with whatever they get as long as it looks decent. Linux users tend to be much worse in this regard, as those are usually the guys who do insist that every application look the same. Linux users are actually the main reason why I kept the classic option to begin with. If it were only macOS, I might have removed it entirely, not least since classic on macOS looks the worst of all platforms.

pigsflew commented:
Why not use [Google's material design icons](https://material.io/resources/icons/?style=baseline) directly? They're all available under Apache 2.0 and available from the linked website as SVG.

I've already pulled a bunch of them to set as my icons as external ones, and they look like this:
![example](https://user-images.githubusercontent.com/868821/83212799-58894e00-a115-11ea-9203-d0ae7e7448fe.png)

michaelblyons commented:
The 2.6.0 went out on Homebrew recently, and I'm sure you're getting a flurry of issue reports about one thing or another...

But I feel compelled to say these themes are _amazing_. Great work!

phoerious commented:
Appreciated.

zyv commented:
Hey, as a Mac person, I was suffering from the native theme being horrible and even more horrible in the dark mode, but I kept suffering because KeePassXC is just too good.

Now, I don't need to suffer anymore. I'm amazed by the greatness of the new theme after I installed the update. It looks absolutely brilliant in the dark mode, and blends a lot better with the native environment than the old one ever did. I haven't used it extensively, but the first impression was "WOW!!!11" - even if there are small bugs here & there, it's just such a huge leap forward as compared to the old theme, and I'm sure with Big Sur it will only get better in terms of blending into the environment.

Everyone who has worked on making this a reality, huge thanks to all of you, you work is greatly appreciated!

zyv commented:
Probably the first macOS-specific bug ;) the separator is rendered as a box with stripes:

<img width="347" alt="Screenshot 2020-07-11 at 10 56 08" src="https://user-images.githubusercontent.com/607401/87220591-64e40600-c365-11ea-93a9-f5d7f6ee09d9.png">

phoerious commented:
Could you make a separate bug report for that please? Looks like your font has a different understanding og what bullet points are supposed to look like.

kattjevfel commented:
Well this certainly looks terrible out of place now, can we get a version with system colours please? Or the very least an option to bring the old theme back?

Application to the right is using native dark theme colours, like it should.
![](https://i.imgur.com/JpvFUQA.png)

The-Compiler commented:
If anyone else is looking for the option @kattjevfel mentioned: It's in the menu under "View -> Theme -> Classic". I also expected to find this in the UI options and ended up searching in the source code to find the menu entry :wink:

droidmonkey commented:
Should have searched the documentation :stuck_out_tongue:

The-Compiler commented:
@droidmonkey Fair point, thanks! Note to self: Only because the documentation for [my own open-source project](https://www.qutebrowser.org/) is rather terse, don't assume that's the case for things I use as well :laughing: Props for the nice documentation with topics and screenshots and all!

Kadah commented:
Just wanted to note that "Classic" only affects the color scheme, the name of the option gave the impression that it would possibly do more. The flat Modern UI does not agree with my eyes, the text is difficult to read (something about Modern UI causes my vision to blur over time). Not saying that is KeepPassXC's issue, I lay that blame on W10 and the Modern UI design spec, the whole OS in defaults does this to me too.

If anything was to change about this... "Classic" could be labeled "Use system colors", or just a hover tip of the latter.

I'm a edge case, so for W10 I'll stick with 2.5.4.

droidmonkey commented:
2.5.4 theming and classic themeing are exactly the same thing....

Kadah commented:
> 2.5.4 themeing and classic themeing in 2.6 are exactly the same thing....

Went back to double check. You're partly correct, the text rendering is actually the same in classic+compact on 2.6.2. 

Would seem its just the new flat blurry iconography combined with W10 UI that is triggering the problem and makes the whole window painful to interact with. On an old W7 install in classic but not compact, the new icons aren't quite as troublesome, but still too harsh to be usable for me.
